File: find_english.py
import os
import re
from bs4 import BeautifulSoup

def is_english_text(text):
    # Strip whitespace and common punctuation/numbers
    text = re.sub(r'[\d\W_]+', '', text)
    # Check if there are any English characters
    if re.search(r'[a-zA-Z]', text):
        return True
    return False

def check_html_files(directory):
    for root, _, files in os.walk(directory):
        if 'node_modules' in root or 'dist' in root:
            continue
        for file in files:
            if file.endswith('.html'):
                path = os.path.join(root, file)
                with open(path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                soup = BeautifulSoup(content, 'html.parser')
                # Remove scripts and styles
                for script in soup(["script", "style", "svg"]):
                    script.extract()
                
                for element in soup.find_all(string=True):
                    text = element.strip()
                    if len(text) > 2 and is_english_text(text):
                        # Filter out common tags/attributes that bs4 might accidentally extract, or known safe English terms if any (like Nammerha)
                        if "{" in text or "}" in text: continue # probably inline JS or Vue/Alpine templates
                        print(f"{path}: Found English text -> {text[:100]}")

if __name__ == '__main__':
    check_html_files('./frontend')

File: find_english.js
const fs = require('fs');
const path = require('path');

function isEnglishText(text) {
    const cleaned = text.replace(/<[^>]+>/g, ' ')
                       .replace(/&[a-z]+;/g, ' ')
                       .replace(/[0-9\W_]+/g, ' ')
                       .trim();
    return /[a-zA-Z]{4,}/.test(cleaned); // looking for words 4+ chars
}

const safeWords = new Set(['html', 'head', 'body', 'script', 'style', 'div', 'span', 'class', 'id', 'meta', 'link', 'href', 'src', 'alt', 'nammerha', 'phosphor', 'icons', 'vite', 'module', 'true', 'false', 'null', 'undefined', 'async', 'defer', 'charset', 'utf', 'viewport', 'content']);

function findEnglishInHTML(dir) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const fullPath = path.join(dir, file);
        if (fs.statSync(fullPath).isDirectory()) {
            if (!fullPath.includes('node_modules') && !fullPath.includes('dist')) {
                findEnglishInHTML(fullPath);
            }
        } else if (fullPath.endsWith('.html')) {
            const content = fs.readFileSync(fullPath, 'utf-8');
            
            // Remove scripts, styles, SVG
            const noScripts = content.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
                                     .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, '')
                                     .replace(/<svg\b[^<]*(?:(?!<\/svg>)<[^<]*)*<\/svg>/gi, '');

            // Find text nodes (stuff between > and <)
            const matches = noScripts.match(/>([^<]+)</g);
            if (matches) {
                matches.forEach(m => {
                    const text = m.substring(1, m.length - 1).trim();
                    if (text.length > 3 && isEnglishText(text)) {
                        const words = text.match(/[a-zA-Z]+/g) || [];
                        const unknownWords = words.filter(w => !safeWords.has(w.toLowerCase()));
                        if (unknownWords.length > 0) {
                            console.log(`${fullPath}: Found -> "${text}" (words: ${unknownWords.join(', ')})`);
                        }
                    }
                });
            }
            
            // Check for missing RTL or hardcoded ltr
            if (content.includes('dir="ltr"')) {
                console.log(`${fullPath}: Found dir="ltr"! Should be RTL.`);
            }
            if (!content.includes('dir="rtl"')) {
                console.log(`${fullPath}: Missing dir="rtl" attribute!`);
            }
        }
    }
}

findEnglishInHTML('./frontend');
